core/common/src/main/java/alluxio/master/metastore/rocks/RocksUtils.java

```diff
@@ -94,35 +99,60 @@ public static long readLong(byte[] bytes, int start) {
    * Used to wrap an {@link CloseableIterator} over {@link RocksIterator}.
    * It seeks given iterator to first entry before returning the iterator.
    *
+   * The abort check is checked in hasNext(), where we check whether the RocksDB is closed and
+   * iteration should end. However, hasNext() and next() is a check-then-act race condition,
+   * where RocksDB may be closed immediately after hasNext() and before next(). We avoid that by
+   * acquiring a lock in next() access. The lock should guarantee safety during the data access.
+   *
+   * With the thread safety baked into hasNext() and next(), users of this Iterator do not need
+   * to worry about safety and can use this Iterator normally.
+   * See examples in how this iterator is used in RocksBlockMetaStore and RocksInodeStore.
+   *
    * @param rocksIterator the rocks iterator
    * @param parser parser to produce iterated values from rocks key-value
    * @param <T> iterator value type
+   * @param abortCheck if true, abort the iteration
+   * @param locker acquire a lock in next() to ensure thread safety, if any
    * @return wrapped iterator
    */
   public static <T> CloseableIterator<T> createCloseableIterator(
-      RocksIterator rocksIterator, RocksIteratorParser<T> parser) {
+      RocksIterator rocksIterator, RocksIteratorParser<T> parser,
+      Supplier<Boolean> abortCheck, Supplier<LockResource> locker) {
     rocksIterator.seekToFirst();
     AtomicBoolean valid = new AtomicBoolean(true);
     Iterator<T> iter = new Iterator<T>() {
       @Override
       public boolean hasNext() {
-        return valid.get() && rocksIterator.isValid();
+        /*
+         * There can be a race condition where after passing the hasNext() check, the RocksDB
+         * is closed and next() causes segfault. Avoiding that requires the closeCheck to stop the
+         * iteration way before entering the critical section. That means we can use the
+         * closeCheck to stop iteration when the iterator is still valid, then safely close
+         * RocksDB and all relevant references without worrying about concurrent readers
+         * like this escaped iterator.
+         */
+        return (!abortCheck.get()) && valid.get() && rocksIterator.isValid();
       }
 
       @Override
       public T next() {
-        try {
-          return parser.next(rocksIterator);
+        boolean succeeded = false;
+        // This lock will assure the RocksDB is not closed while this is reading
+        try (LockResource lock = locker.get()){
+          T result = parser.next(rocksIterator);
+          rocksIterator.next();
+          succeeded = true;
+          return result;
         } catch (Exception exc) {
           LOG.warn("Iteration aborted because of error", exc);
-          rocksIterator.close();
-          valid.set(false);
           throw new RuntimeException(exc);
+        } catch (Throwable t) {
+          LOG.warn("Iteration aborted because of error", t);
```

Comment:
if this log is not that necessary, we can just remove this catch, assuming the uncaught ex will be logged later anyway